File: JavaScript - C#/6- Singly Linked List/LinkedList - JS/linkedList.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.length = 0;
    this.head = null;
    this.tail = null;
  }

  // Add a new node at the end of the linked list.
  // O(1)
  push(val) {
    var node = new Node(val);

    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = node;
    }

    this.length++;
    return this;
  }

  // Removes the tail element
  // O(n)
  pop() {
    if (!this.head) return null;

    var curr = this.head;
    var newTail = this.head;

    while (curr.next) {
      // while my current has a next, new tail should be the one before that
      // until you reach the end
      newTail = curr;
      curr = curr.next;
    }

    this.tail = newTail;
    this.tail.next = null;
    this.length--;

    if (this.length === 0) {
      this.head = null;
      this.tail = null;
    }

    return curr;
  }

  // Removes an element from the beginning
  // O(1)
  shift() {
    if (!this.head) return null;

    var currHead = this.head;
    this.head = currHead.next;
    this.length--;

    if (this.length === 0) this.tail = null;

    return currHead;
  }

  // Adds an element to the beginning
  // O(1)
  unshift(val) {
    var newNode = new Node(val);

    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head = newNode;
    }

    this.length++;
    return this;
  }

  // Returns a node from the given index
  // O(n)
  get(index) {
    if (index < 0 || index >= this.length) return null;

    var counter = 0;
    var curr = this.head;

    while (counter !== index) {
      curr = curr.next;
      counter++;
    }

    return curr;
  }

  // Sets the value of a given index
  // O(n)
  set(index, val) {
    var foundNode = this.get(index);
    if (foundNode) {
      foundNode.val = val;
      return true;
    }
  }

  // Inserts a value into a given index
  // O(n)
  insert(index, val) {
    if (index < 0 || index > this.length) return null;
    else if (index === this.length) this.push(val);
    else if (index === 0) this.unshift(val);
    else {
      var prev = this.get(index - 1);
      var newNode = new Node(val);

      newNode.next = prev.next;
      prev.next = newNode;

      this.length++;
      return this.get(index);
    }
  }

  // Removes a node from a given index
  // O(n)
  remove(index) {
    if (index < 0 || index >= this.length) return null;
    if (index === 0) return this.shift();
    if (index === this.length - 1) return this.pop();

    var prev = this.get(index - 1);
    var removed = prev.next;

    prev.next = removed.next;
    this.length--;

    return removed;
  }

  // Reverses the linked list
  // O(n)
  reverse() {
    var curr = this.head;
    this.head = this.tail;
    this.tail = curr;

    var prev = null;
    var next = null;

    for (var i = 0; i < this.length; i++) {
      next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }

    return this;
  }

  // Helper print function to visualize the Linked list
  // O(n)
  print(message) {
    var curr = this.head;
    while (curr) {
      process.stdout.write(`${curr.val} --> `);
      curr = curr.next;
    }
    console.log(`null \nCurrent state of the linked list after ${message}`);
    console.log(`Length: ${this.length}\n`);
  }
}

var linkedList = new SinglyLinkedList();
linkedList.push("Onur");
linkedList.push("Ucar");
linkedList.push("Hello");
linkedList.push("!");
linkedList.print("4 pushes");

linkedList.pop();
linkedList.print("pop");

linkedList.unshift("Unshift");
linkedList.print("unshift");

linkedList.shift();
linkedList.print("shift");

linkedList.set(2, "REPLACED");
linkedList.print("set at index 2");

linkedList.insert(3, "World");
linkedList.print("insert at index 3");

linkedList.remove(1);
linkedList.print("removing index 1");

linkedList.reverse();
linkedList.print("reversing the linked list");

console.log(
  `Head: ${linkedList.head.val}\nTail: ${linkedList.tail.val}\nLength: ${linkedList.length}`
);

File: JavaScript - C#/6- Singly Linked List/LinkedList - CS/LinkedList.cs
using System;
namespace Playground
{
	public class LinkedList
	{
		private Node? _Head { get; set; }

		private Node? _Tail { get; set; }

		private int _Length { get; set; }


		public LinkedList()
		{
			_Head = null;
			_Tail = null;
			_Length = 0;
		}

		// Helper Function(s)
		public bool IsEmpty()
		{
			return (_Head == null) ? true : false;
		}

		public void Print()
		{
			Node curr = _Head;

			while (curr != null)
			{
				Console.Write($"{curr._Data} --> ");
				curr = curr._Next;
			}

			Console.WriteLine("null");
			Console.WriteLine($"Length: {_Length}\n");
		}



		// Push: Add a new node at the end of the linked list
		// O(n)
		public Node Push(int val)
		{
			Node node = new Node(val);

			if (IsEmpty())
			{
				_Head = node;
				_Tail = node;
			} else
			{
				_Tail._Next = node;
				_Tail = node;
			}

			_Length++;

			return _Tail;
		}

		// Pop: Removes the tail element
		// O(n)
		public Node Pop()
		{
			if (IsEmpty()) return null;

			Node current = _Head;
			Node newTail = _Head;

			while (current._Next != null)
			{
				newTail = current;
				current = current._Next;
			}

			_Tail = newTail;
			_Tail._Next = null;
			_Length--;

			if(_Length == 0)
			{
				_Tail = null;
				_Head = null;
			}

			return newTail;

		} 

		// Shift: Removes an element from the beginning
		// O(1)
		public Node Shift()
		{
			if (IsEmpty()) return null;

			Node currentHead = _Head;

			// setting the new head
			_Head = currentHead._Next;
			_Length--;

			if(_Length == 0)
			{
				_Tail = null;
			}

			return _Head;
		}

		// Unshift: Adds an element to the beginning
		// O(1)
		public Node Unshift(int val)
		{
			Node node = new Node(val);

			if (IsEmpty())
			{
				_Head = node;
				_Tail = node;
			} else
			{
				node._Next = _Head;
				_Head = node;
			}

			_Length++;

			return _Head;
		}

        // Get: Returns a node from a given index
		// O(n)
		public Node Get(int index)
		{
			if (index < 0 || index >= _Length) return null;

			int counter = 0;
			Node current = _Head;

			while (counter != index)
			{
				current = current._Next;
				counter++;
			}

			return current;
		}


        // Set: Sets the value of a given index
		// O(n)
		public Node Set(int index, int val)
		{
			Node node = Get(index);

			node._Data = val;

			return node;
		}


		// Insert: Inserts a value into a given index
		// O(n)
		public Node Insert(int index, int val)
		{
			if (index < 0 || index > _Length) return null;

			else if (index == 0) return Unshift(val);

			else
			{
                Node newNode = new Node(val);

                Node previous = Get(index - 1);

                newNode._Next = previous._Next;

                previous._Next = newNode;

                _Length++;

                return newNode;
            }
		}


		// Remove: Removes a node from a given index
		// O(n)
		public Node Remove(int index)
		{
			if (index < 0 || index >= _Length) return null;

			else if (index == 0) return Shift();

			else if (index == _Length - 1) return Pop();

			else
			{
				Node previous = Get(index - 1);
				Node removedNode = previous._Next;

				previous._Next = removedNode._Next;
				removedNode._Next = null;

				this._Length--;

				return removedNode;
			}
		}


		// !!!
		// Reverse
		// O(n)
		public void Reverse()
		{
			Node current = _Head;
			_Head = _Tail;
			_Tail = current;

			Node prev = null;
			Node next = null;

			for(int i = 0; i < _Length; i++)
			{
				next = current._Next;
				current._Next = prev;
				prev = current;
				current = next;
			}
		}

    }
}

